<!DOCTYPE html>
<html>

<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <title>CS 241 - Preparation Material</title>
    <link rel="stylesheet" type="text/css" href="../course/style.css" />
</head>

<body>
<div id="courseTitle">
    <span class="icon-byui-logo"></span>
    <h1>OO Programming and Data Structures | CS 241</h1>
</div>
    <article>
        <h2>09 Prepare : Reading</h2>


        <h3>Outcomes</h3>
        <p>At the end of this week, successful students will be able to:</p>
        <ol>
            <li><p>Show fluency in discussing exceptions and error handling.</p></li>
            <li><p>Write programs that correctly use exceptions and error handling to solve problems.</p></li>
        </ol>

        <h3>Preparation Material</h3>
        <p>For this week, please read the following introduction to Exception/Error handling:</p>

        <ul>
            <li><p><a href="https://www.python-course.eu/exception_handling.php" target="_blank">Exception Handling</a></p></li>
        </ul>

        <h4>A note about one of the examples in the reading:</h4>
        <p>For one of the examples in the reading, they showed how to use a <code>while True</code> loop with a <code>break</code> statement. While this works, it can sometimes make the code difficult to follow to have loops exit at non-standard times (i.e., using break statements), so it is usually considered better practice to let the loop's condition cause it to exit.</p>
        <p>With this in mind, we could rework the code from the example to be something more like:</p>

<pre><code class="python">
valid_input = False

while not valid_input:
    try:
        n = int(input("Please enter an integer: "))
        valid_input = True
    except ValueError:
        print("Not a valid integer, please try again.")

print("Successfully entered an integer.")
</code></pre>

        <h4>Creating Custom Types of Errors</h4>
        <p>In addition to using the built-in types, we can also define our own custom types of errors. See the <a href="https://docs.python.org/3/tutorial/errors.html#user-defined-exceptions" target="_blank">official Python documentation</a> for an example of how to do this, but in short, assuming we wanted to create an exception class for a InvalidStudentError, the Python code would look as follows:</p>

<pre><code class="python">
class InvalidStudentError(Exception):
    def __init__(self, message):
        # Don't forget to pass the message to the base class
        super().__init__(message)

</code></pre>

Then we could use our new InvalidStudentError, just like any other error:

<pre><code class="python">
if student.name == None:
    raise InvalidStudentError("The student must have a name")
</code></pre>


       <script src="../course/js/highlight/highlight.pack.js"></script>
       <script>hljs.initHighlightingOnLoad();</script>


    </article>
</body>

</html>